core/src/main/java/io/hyperfoil/core/impl/AffinityAwareSessionPool.java

```diff
@@ -0,0 +1,271 @@
+        }
+    }
+
+    private void decrementUsed() {
+        var counters = this.counters;
+        int used = counters.decrementAndGet(USED_OFFSET);
+        int minUsed = counters.get(MIN_USED_OFFSET);
+        if (minUsed > 0 && used < minUsed) {
+            counters.lazySet(MIN_USED_OFFSET, used);
+        }
+    }
+
+    @Override
+    public void release(Session session) {
+        Objects.requireNonNull(session);
+        var localQueue = localQueues[session.agentThreadId()];
+        decrementUsed();
+        localQueue.add(session);
+    }
+
+    @Override
+    public void reserve(int capacity) {
+        int totalCapacity = getLocalQueuesCapacity();
+        if (totalCapacity >= capacity) {
+            return;
+        }
+        moveNewSessionsToLocalQueues(createSessionShards(capacity, totalCapacity));
+    }
+
+    private int getLocalQueuesCapacity() {
+        int totalCapacity = 0;
+        for (Queue<Session> localQueue : localQueues) {
+            if (localQueue != null) {
+                totalCapacity += localQueue.size();
+            }
+        }
+        return totalCapacity;
+    }
+
+    private void moveNewSessionsToLocalQueues(Queue<Session>[] configLocalQueues) {
+        for (int i = 0; i < configLocalQueues.length; i++) {
+            var newSessions = configLocalQueues[i];
+            var runtimeLocalQueue = localQueues[i];
+            if (runtimeLocalQueue == null) {
+                runtimeLocalQueue = createLocalQueue(newSessions.size());
+            } else {
+                var newLocalQueue = createLocalQueue(newSessions.size() + runtimeLocalQueue.size());
+                newLocalQueue.addAll(runtimeLocalQueue);
+                runtimeLocalQueue.clear();
+                runtimeLocalQueue = newLocalQueue;
+            }
+            runtimeLocalQueue.addAll(newSessions);
+            newSessions.clear();
+            localQueues[i] = runtimeLocalQueue;
+        }
+    }
+
+    private Queue<Session>[] createSessionShards(int requiredCapacity, int availableCapacity) {
+        var newSessionsPerEventExecutors = new Queue[localQueues.length];
+        var additionalCapacity = requiredCapacity - availableCapacity;
+        var perEventExecutorCapacity = (int) Math.ceil((double) additionalCapacity / localQueues.length);
+        for (int i = 0; i < additionalCapacity; i++) {
+            var newSession = sessionSupplier.get();
+            if (newSession == null) {
+                throw new IllegalStateException("Session supplier returned null");
+            }
+            var executor = newSession.executor();
+            var boxedAgentThreadId = agentThreadIdPerExecutor.get(executor);
+            if (boxedAgentThreadId == null) {
+                throw new IllegalStateException("This session is associated with an unknown EventExecutor");
+            }
+            // sanity check against the sessions agentThreadId
+            if (boxedAgentThreadId.intValue() != newSession.agentThreadId()) {
```

Comment:
```suggestion
            if (boxedAgentThreadId != newSession.agentThreadId()) {
```

Reply:
I don't remember by heart if the left side in unboxed or the right side is boxed, so I preferred to be safe and avoid a potential allocation path.